### PR TITLE
[substrait] Add support for ExtensionTable

### LINKED ISSUE #13771: [substrait] Add support for ExtensionTable
### Is your feature request related to a problem or challenge?

Custom TableProvider implementations cannot currently be encoded as ExtensionTables in Substrait. The Substrait plan will only retain the table names, which are hardly ever enough to restore the table definition on the consumer side. For UDTFs in particular, the name is completely useless, as it's always `tmp_table`.

### Describe the solution you'd like

Add two more methods in `SerializerRegistry` for serializing/deserializing `TableSource` instances and use these new extensions in `to_substrait_plan` and `from_substrait_plan` to let users encode/decode custom table definitions as ExtensionTables.

### Describe alternatives you've considered

For the cases where the user controls the table name, one hideous workaround would be to encode the whole table definition in the table name and register a custom schema provider to decode it (e.g. `some_catalog.custom_schema."base64(proto_binary)"`). This is a horrible hack as it requires using those names in SQL queries and it doesn't work for UDTFs.
A far better alternative is to leverage the already supported Substrait extensions (in particular, `ExtensionLeaf`), by implementing the `SerializerRegistry` trait and forcing the table to fit into a `UserDefinedLogicalNode`s.
However, this approach is both limited and unnatural:
- logical plans have to be preprocessed in order to replace `TableScan`s with `Extension` nodes before converting to substrait
- The logical plan resulting from decoding Substrait can only be executed if an `ExtensionPlanner` is registered for handling the user-defined nodes, but in this case it would not benefit from the special treatment that tables get in DataFusion (projections and filter pushdowns for scan, various knobs to instruct the engine about the table capabilities etc.). Rewriting the decoded plan to convert the user-defined node back to a TableScan is the only way to benefit from all that.
- Substrait also encodes projections only for tables (i.e. `ReadRel`s), so an ExtensionLeaf can't make use of that. Rewriting the Substrait plan itself to overcome this limitation is way more tedious than it should be.

### Additional context

This should be a child issue of https://github.com/apache/datafusion/issues/13318.

### PR DESCRIPTION
**Pull request by ccciudatu**
## Which issue does this PR close?

Closes #13771.
Addresses the first bullet in #13318.

## Rationale for this change

Adds support for encoding/decoding custom table providers as `ExtensionTable`s in Substrait.

## What changes are included in this PR?

Two more methods for `SerializerRegistry` to handle `TableSource`s and the necessary changes in `from/to _substrait_plan` to transparently map custom tables to ExtensionTable nodes.

## Are these changes tested?

A round trip test is included.

## Are there any user-facing changes?

No breaking changes, only a couple of convenience changes, such as default implementations for trait methods.

### CODE REVIEW

File: datafusion/core/src/execution/context/mod.rs [hunk 1]
```diff
@@ -1679,27 +1679,7 @@ pub enum RegisterFunction {
...
-    fn deserialize_logical_plan(
-        &self,
-        name: &str,
-        _bytes: &[u8],
-    ) -> Result<Arc<dyn UserDefinedLogicalNode>> {
-        not_impl_err!(
-            "Deserializing user defined logical plan node `{name}` is not supported"
-        )
-    }
-}
```
Comment (ccciudatu) [inline datafusion/core/src/execution/context/mod.rs:1702]: The `SerializerRegistry` trait now has two more methods for handling tables (with default implementations for backwards compatibility), so it makes sense for the existing methods to have default implementations as well.
This will allow implementors to conveniently implement the trait for user-defined logical nodes only or for tables only.
Since the implementations here are perfect as trait defaults, this PR just moves them into the trait itself.

File: datafusion/substrait/src/logical_plan/consumer.rs [hunk 2]
```diff
@@ -994,8 +994,34 @@ pub async fn from_substrait_rel(
                     )
                     .await
                 }
-                _ => {
-                    not_impl_err!("Unsupported ReadType: {:?}", &read.as_ref().read_type)
+                Some(ReadType::ExtensionTable(ext)) => {
```
Comment (vbarua) [inline datafusion/substrait/src/logical_plan/consumer.rs:997]: I didn't actually get to handling ExtensionTables in https://github.com/apache/datafusion/pull/13803, but I think I've got most what we would need in place for it.

The way I would approach it, based on the work in that PR, is to add a method to the `SubstraitConsumer` trait like:
```rust
async fn consume_extension_table(&self, extension_table: &ExtensionTable) -> Result<LogicalPlan>
```
and wire it in here. Then, as a user you would be able to provide your own implementation of the decoder. This _might_ user the SerializerRegistry, but it doesn't necessarily need to.

Comment (ccciudatu) [reply datafusion/substrait/src/logical_plan/consumer.rs:997]: That's great! I like where that is going.
My goal here was to add the missing support for reading & writing extension tables leveraging only what's available (to keep the patch as small as possible).
But I do agree that uniform handling of Substrait extensions would make more sense and would overcome some of the current limitations.

I think the code here is really easy to migrate to the new `SubstraitConsumer` (and `SubstraitProducer`?) interfaces once they're available, by just replacing the `SerializerRegistry` calls with the new dedicated read/write methods. But the rest would be mostly unchanged by this migration.
FWIW, once your refactoring is complete, I think there would be no place for `SerializerRegistry` anymore, and it should be removed (or at least deprecated).

Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs:997]: > (and SubstraitProducer?)

I haven't gotten around to the producer yet, but if the SubstraitConsumer pattern makes sense to folks it should be easy enough to hammer it out.

> I think the code here is really easy to migrate to the new SubstraitConsumer

Would you be open to having https://github.com/apache/datafusion/pull/13803 merged first, and then porting your code over?

Comment (ccciudatu) [reply datafusion/substrait/src/logical_plan/consumer.rs:997]: Sure, I can do that.

**Comment by alamb** [comment]
- I believe the plan is to merge https://github.com/apache/datafusion/pull/13803 first then rework this code, so marking PR as draft to signify it isn't waiting on review

**Comment by ccciudatu** [comment]
@vbarua I rebased on top of https://github.com/apache/datafusion/pull/13803 and added a `consume_extension_table` method to the `SubstraitConsumer` trait.
I chose to pass the schema and projection to this method (instead of keeping the `TableScan` postprocessing in `from_read_rel`) to allow custom implementations to use that information for fully restoring their custom tables if needed.

However, there are a few things that I'd like to revisit once we have a (symmetrical) customisable `SubstraitProducer` alongside the consumer. In particular, I want to move away from the `SerializerRegistry` trait and provide full control to the user over the serialisation of the whole `ReadRel` in case of extension tables by working with proto objects directly instead of  trying to model this as a generic serde concern. This would allow the user to:
1. store the table name in (e.g.) `ReadRel::common::hint::alias` instead of abusing the `proto::Any::type_url` field of `ExtensionTable::detail`
2. Use the `ExtensionTable::detail::type_url` field to distinguish between multiple table implementations (as per the protobuf spec)
3. leverage the available ReadRel information that is currently unused (e.g. filtering, advanced extensions etc.)

File: datafusion/substrait/src/logical_plan/consumer.rs [hunk 3]
```diff
@@ -438,6 +439,22 @@ pub trait SubstraitConsumer: Send + Sync + Sized {
             user_defined_literal.type_reference
         )
     }
+
+    fn consume_extension_table(
+        &self,
+        extension_table: &ExtensionTable,
+        _schema: &DFSchema,
+        _projection: &Option<MaskExpression>,
+    ) -> Result<LogicalPlan> {
```
Comment (vbarua) [inline datafusion/substrait/src/logical_plan/consumer.rs:448]: > I chose to pass the schema and projection to this method (instead of keeping the TableScan postprocessing in from_read_rel) to allow custom implementations to use that information for fully restoring their custom tables if needed.

I like where your head is at with this, but I almost want to go further. You already called out:
> leverage the available ReadRel information that is currently unused (e.g. filtering, advanced extensions etc.)

Maybe the interface for this should just be:
```rust
    fn consume_extension_table(
        &self,
        read_rel: &ReadRel
        extension_table: &ExtensionTable) -> Result<LogicalPlan>
```
which will be future proofed for if fields are ever added to the ReadRel, and also provides access to common fields on the ReadRel.

We could even go further and add
```
    fn consume_named_table(
        &self,
        read_rel: &ReadRel
        named_table: &NamedTable) -> Result<LogicalPlan>

    fn consume_virtual_table(
        &self,
        read_rel: &ReadRel
        named_table: &VirtualTable) -> Result<LogicalPlan>
```

to make it easier to customize behaviour for specific read_types. This last idea might be better as it's own PR, as we would need to factor out some of the code in `from_read_rel` into functions to be re-used across the new helpers.

**Comment by vbarua** [comment]
Left some feedback about the `consume_extension_table` API.

> However, there are a few things that I'd like to revisit once we have a (symmetrical) customisable SubstraitProducer alongside the consumer.

I'm planning on working that over the holidays https://github.com/apache/datafusion/issues/13901, so that might come round faster than you think.

> I want to move away from the SerializerRegistry trait and provide full control to the user over the serialisation of the whole ReadRel in case of extension tables by working with proto objects directly instead of trying to model this as a generic serde concern.

100%

> store the table name in (e.g.) ReadRel::common::hint::alias instead of abusing the proto::Any::type_url field of ExtensionTable::detail

Honestly, if a user was doing this I would recommend that they store the name in the extension message directly.

File: datafusion/substrait/src/logical_plan/consumer.rs (hunk 3 above)
Comment (ccciudatu) [reply datafusion/substrait/src/logical_plan/consumer.rs:448]: I can see your point, especially as I've been toying with the same idea myself.
However, I found the intrinsic redundancy a bit error prone: at least in theory, this interface allows the `ReadRel` to contain a table other than the one passed in as the second argument.
Eliminating this redundancy would end up with the exact same signature of `consume_read`, which renders the new helper(s) superfluous.

**Comment by ccciudatu** [comment]
> I'm planning on working that over the holidays #13901, so that might come round faster than you think.

@vbarua That's great, especially since this PR feels a bit half-baked after the last rebase (as an attempt to provide a coherent roundtrip between the current inflexible producer and the new `SubstraitConsumer`).

Maybe I should just hold this until the `SubstraitProducer` is added and reassess whether this extension point is still needed or this capability can be achieved by simply providing custom trait implementations with fallbacks for `produce_read` and `consume_read` respectively, thus requiring no explicit support in the library code.

File: datafusion/substrait/src/logical_plan/consumer.rs (hunk 3 above)
Comment (ccciudatu) [reply datafusion/substrait/src/logical_plan/consumer.rs:448]: @Blizzara You are most likely right. My main concern was to make sure the table name gets propagated for extension tables as well.
As @vbarua suggested, the name can be encoded in the extension payload, but I found that redundant, possibly because I misread the Substrait spec and assumed the name already had its reserved place outside the ExtensionTable::detail, but I'm a lot less convinced about that now.

I'll have a look at the new substrait producer and give it another try, unless this feature can be fully covered by implenting the Producer/Consumer traits, in which case I'll probably just close this PR without merging.

**Comment by ccciudatu** [comment]
@vbarua @Blizzara I finally got back to figure out whether it still makes sense with the new APIs.
With the latest changes, supporting extension tables no longer requires a fork of datafusion, but merely a custom implementation for the new traits. However, there's a bit of boilerplate/delegation/copy-paste required for leveraging the default functionality.

So I decided to give this another try with a few changes in a fixup commit:

- I changed the `SerializerRegistry` trait to return a string qualifier alongside the bytes on serialize. This way, the implementor can make proper use of the protobuf `type_url` or can use any string qualifier as a discriminator for the actual TableProvider implementation.
- Since the ExtensionTable::detail::type_url is now supposed to be the proper protobuf one, I chose to store the original table name as `rel.common.hint.alias`, hoping to convince @vbarua that a custom TableProvider implementation is hardly aware of the names under which users choose to register the corresponding tables (especially for UDTFs), so it's not trivial to make the name part of the extension payload and it imposes unnecessary restrictions for the supported table providers. The alias property is loosely specified; the canonical example seems to be the SubqueryAlias, but it's not limited to that and it feels "close enough". A default name will be used if the hint is missing.

**Comment by vbarua** [comment]
> With the latest changes, supporting extension tables no longer requires a fork of datafusion, but merely a custom implementation for the new traits.

Nice!

> However, there's a bit of boilerplate/delegation/copy-paste required for leveraging the default functionality.

@ccciudatu do you have an example of any standard and/or built-in DataFusion functionality that can't be serialised to Substrait without this boilerplate? Is it primarily or only UserDefineTypeRels?

I'm realising that I don't have a good idea of what this would actually be needed for in standard DataFusion. If there is existing functionality that needs this kind of handling to work, I think it makes sense to include it in the default consumer and producer.

Most of my experience with ExtensionTables in Substrait (and hence biases) comes from generating plans outside of DataFusion where we have table-like sources, like embedding a SQL query for another system or fetching data from an internal cache store, which aren't part of standard Substrait (and are also highly-specific to our use case and deployments). For stuff like this we define our own handling because we don't expect DataFusion to be able to handle our custom messages.

> a custom TableProvider implementation is hardly aware of the names under which users choose to register the corresponding tables (especially for UDTFs)

That's somewhat my argument for not supporting custom stuff generally. The only people who can know what names they used are the ones generating the plans and wiring in the UDTFs in the first place. As long as they have the hooks in place to wire things up how they want, we can service their needs.

**Comment by ccciudatu** [comment]
Thanks, @vbarua. And sorry about the delay.


> @ccciudatu do you have an example of any standard and/or built-in DataFusion functionality that can't be serialised to Substrait without this boilerplate?

No, I don't think any standard/built-in component is ever supposed to end up as an `ExtensionTable` or `ExtensionRel` of any kind.
However, since we do have explicit support for `UserDefinedNodeType`s, it kind of makes sense to add support for user-defined tables as well. Trying to represent custom tables as `ExtensionLeafRel`s in Substrait and `UserDefinedNodeType`s has several limitations that I hinted at in the associated issue, e.g. the table context information gets lost in an `ExtensionLeafRel` as opposed to a `ReadRel`, `ExtensionPlanner::plan_extension` is not a suitable wrapper around `TableProvider::scan`, both producers and consumers need to pre/post-process the logical plans before/after conversion etc.

> Is it primarily or only UserDefineTypeRels?

Yes, this is only meant to address user-defined / custom `TableProvider` implementations, and only the ones that cannot be restored by name alone -- otherwise a `NamedTable` read_type would suffice.


> I'm realising that I don't have a good idea of what this would actually be needed for in standard DataFusion. If there is existing functionality that needs this kind of handling to work, I think it makes sense to include it in the default consumer and producer.

Again, I'm not sure if you consider extension points like `UserDefinedLogicalNode` and `TableProvider` as part of "standard DataFusion".  However, I see them as equally legit, i.e. if there's room for the first in the default producer/consumer, the latter should also be achievable with just a custom serializer.

> Most of my experience with ExtensionTables in Substrait (and hence biases) comes from generating plans outside of DataFusion where we have table-like sources, like embedding a SQL query for another system or fetching data from an internal cache store, which aren't part of standard Substrait (and are also highly-specific to our use case and deployments). For stuff like this we define our own handling because we don't expect DataFusion to be able to handle our custom messages.

One use-case that I'm trying to address with this is a UDTF that integrates a legacy OLAP system that we're using as a data source for both DataFusion and Spark, using Substrait to federate the query plan as we see fit.

An SQL query like `select * from legacy_olap_system(arg1, arg2, ...)` results in a `TableScan` node that references a custom `TableProvider` implementation and has the name "tmp_table" assigned by DataFusion.
Encoding/decoding such a plan to/from Substrait with the current producer/consumer results in the equivalent of `select * from tmp_table` and errors out with a "table not found" message.

With the code in this patch and a custom `SerializerRegistry`, such a custom table will encode its internal state as proto and will get fully restored in a different DataFusion context (or a different system altogether).

> > a custom TableProvider implementation is hardly aware of the names under which users choose to register the corresponding tables (especially for UDTFs)
> 
> That's somewhat my argument for not supporting custom stuff generally. The only people who can know what names they used are the ones generating the plans and wiring in the UDTFs in the first place. As long as they have the hooks in place to wire things up how they want, we can service their needs.

I agree this doesn't feel exactly right. The Substrait spec for extension tables is quite poorly defined, so to recreate the exact same plan (including names) on roundtrips we do require either small hacks or a more complex interface.
However, this is not a hard requirement and the code accepts any compliant input. I can even remove this hack for now and wait/request for enhancements in the Substrait spec, if this is too ugly.

**Comment by alamb** [comment]
Hi @ccciudatu  and @vbarua  -- I am trying to work down the backlog of PRs in the review queue -- what is the status of this PR? Is it ready for review / inclusion? Are you still working on a design? Something else?

**Comment by vbarua** [comment]
Apologies for the delay, I haven't had the bandwidth to follow up on this (and I still don't tbh).

At this point I'm ambivalent about this capability, but I wouldn't vote against it. Part of this is that I don't understand the use case sufficiently. To @ccciudatu's point:

> the Substrait spec for extension tables is quite poorly defined, so to recreate the exact same plan (including names) on roundtrips we do require either small hacks or a more complex interface.
However, this is not a hard requirement and the code accepts any compliant input. I can even remove this hack for now and wait/request for enhancements in the Substrait spec, if this is too ugly.

I think that Substrait itself could use some improvements around this. I know there has been some discussion around improving [table function](https://github.com/substrait-io/substrait/issues/753) support in the core spec.

My main concern with merging this would be around wanting/needing to make breaking changes in the future once the spec improves. But also we have done that in the past to improve the API so that shouldn't be a blocker.

TLDR: No strong feelings for or against at this moment.

**Comment by ccciudatu** [comment]
Closing this PR as the new Consumer/Producer APIs make the Substrait conversion fully customizable and the explicit support for `ExtensionTable`s can wait at least until they're fully specified.
